tsinfer/formats.py

```diff
@@ -2445,9 +2460,9 @@ def provenances_record(self):
         except KeyError:
             return np.array([], dtype=object)
 
-    @property
+    @functools.cached_property
     def num_samples(self):
```

Comment:
Is it worth documenting this as the number of samples that have not been masked out (equivalent to the total number of samples in the dataset if there is no masking)?